File: typescript/challenge-one/test_endpoints.ts
/**
 * Script de teste para validar os endpoints da API.
 * Rode este script com o servidor ligado (npm start) para ver quais endpoints
 * estao funcionando corretamente e quais estao com problemas.
 *
 * Uso:
 *   npm test
 */

import http from "http";

const BASE_URL = "http://localhost:5050";

const GREEN = "\x1b[92m";
const RED = "\x1b[91m";
const YELLOW = "\x1b[93m";
const RESET = "\x1b[0m";
const BOLD = "\x1b[1m";

function fetch(path: string): Promise<{ data: any; error: string | null }> {
  return new Promise((resolve) => {
    http
      .get(`${BASE_URL}${path}`, (res) => {
        let body = "";
        res.on("data", (chunk) => (body += chunk));
        res.on("end", () => {
          try {
            resolve({ data: JSON.parse(body), error: null });
          } catch {
            resolve({ data: null, error: `Invalid JSON: ${body}` });
          }
        });
      })
      .on("error", (err) => {
        resolve({ data: null, error: `Servidor inacessivel: ${err.message}` });
      });
  });
}

function check(label: string, actual: any, expected: any): boolean {
  if (JSON.stringify(actual) === JSON.stringify(expected)) {
    console.log(`  ${GREEN}PASS${RESET} ${label}`);
    console.log(`         Esperado: ${JSON.stringify(expected)}`);
    return true;
  } else {
    console.log(`  ${RED}FAIL${RESET} ${label}`);
    console.log(`         Esperado: ${JSON.stringify(expected)}`);
    console.log(`         Recebido: ${JSON.stringify(actual)}`);
    return false;
  }
}

function checkApprox(
  label: string,
  actual: any,
  expected: number,
  tolerance = 0.01
): boolean {
  if (actual !== null && actual !== undefined && Math.abs(actual - expected) < tolerance) {
    console.log(`  ${GREEN}PASS${RESET} ${label}`);
    console.log(`         Esperado: ~${expected}`);
    return true;
  } else {
    console.log(`  ${RED}FAIL${RESET} ${label}`);
    console.log(`         Esperado: ~${expected}`);
    console.log(`         Recebido: ${actual}`);
    return false;
  }
}

interface TestResult {
  passed: number;
  failed: number;
}

async function testEndpoint(
  name: string,
  path: string,
  checksFn: (data: any) => boolean[]
): Promise<TestResult> {
  console.log(`\n${BOLD}--- ${name} ---${RESET}`);
  console.log(`    GET ${path}`);

  const { data, error } = await fetch(path);

  if (error) {
    console.log(`  ${RED}ERRO${RESET} Nao foi possivel acessar o endpoint`);
    console.log(`         ${error}`);
    return { passed: 0, failed: 1 };
  }

  let passed = 0;
  let failed = 0;
  for (const result of checksFn(data)) {
    if (result) passed++;
    else failed++;
  }
  return { passed, failed };
}

async function main() {
  console.log(`${BOLD}Verificando se o servidor esta rodando...${RESET}`);

  const { error } = await fetch("/api/revenue");
  if (error && error.includes("inacessivel")) {
    console.log(
      `\n${RED}ERRO: O servidor nao esta acessivel em ${BASE_URL}${RESET}`
    );
    console.log("Certifique-se de que voce rodou: npm start");
    console.log(`Erro: ${error}`);
    process.exit(1);
  }
  console.log(`${GREEN}Servidor OK!${RESET}`);

  let totalPassed = 0;
  let totalFailed = 0;

  // =============================================
  // TESTE 1: Receita total (deve excluir cancelados)
  // =============================================
  let r = await testEndpoint(
    "Teste 1: Receita Total (excluindo pedidos cancelados)",
    "/api/revenue",
    (data) => [checkApprox("total_revenue", data?.total_revenue, 978.1)]
  );
  totalPassed += r.passed;
  totalFailed += r.failed;

  // =============================================
  // TESTE 2: Receita por restaurante
  // =============================================
  r = await testEndpoint(
    "Teste 2: Receita por Restaurante (excluindo cancelados)",
    "/api/revenue/restaurant",
    (data) => [
      checkApprox("Sushi Place", data?.["Sushi Place"], 384.9),
      checkApprox("Burger House", data?.["Burger House"], 202.2),
    ]
  );
  totalPassed += r.passed;
  totalFailed += r.failed;

  // =============================================
  // TESTE 3: Top restaurantes
  // =============================================
  r = await testEndpoint(
    "Teste 3: Top 3 Restaurantes (maior receita primeiro)",
    "/api/restaurants/top",
    (data) => [
      check(
        "Ordem correta",
        data?.top_restaurants?.map((r: any) => r.name),
        ["Sushi Place", "Pizza Roma", "Burger House"]
      ),
      checkApprox(
        "Receita do #1 (Sushi Place)",
        data?.top_restaurants?.[0]?.revenue,
        384.9
      ),
    ]
  );
  totalPassed += r.passed;
  totalFailed += r.failed;

  // =============================================
  // TESTE 4: Produtos populares
  // =============================================
  r = await testEndpoint(
    "Teste 4: Produtos Mais Populares",
    "/api/products/popular",
    (data) => [
      check(
        "Produto #1 e Milkshake (5 unidades)",
        data?.popular_products?.[0]?.product,
        "Milkshake"
      ),
      check(
        "Quantidade do #1",
        data?.popular_products?.[0]?.quantity,
        5
      ),
    ]
  );
  totalPassed += r.passed;
  totalFailed += r.failed;

  // =============================================
  // TESTE 5: Resumo geral
  // =============================================
  r = await testEndpoint(
    "Teste 5: Resumo Geral",
    "/api/summary",
    (data) => [
      check("total_orders (sem cancelados)", data?.total_orders, 12),
      check("cancelled_orders", data?.cancelled_orders, 3),
      checkApprox("total_revenue", data?.total_revenue, 978.1),
      check("total_items_sold", data?.total_items_sold, 28),
    ]
  );
  totalPassed += r.passed;
  totalFailed += r.failed;

  // =============================================
  // RESULTADO FINAL
  // =============================================
  const total = totalPassed + totalFailed;
  console.log(`\n${"=".repeat(50)}`);
  console.log(`${BOLD}RESULTADO FINAL: ${totalPassed}/${total} testes passando${RESET}`);

  if (totalFailed === 0) {
    console.log(`${GREEN}Todos os testes passaram! Parabens!${RESET}`);
  } else {
    console.log(`${RED}${totalFailed} teste(s) falhando${RESET}`);
    console.log(
      `\n${YELLOW}Dica: corrija os bugs um de cada vez e rode este script novamente.${RESET}`
    );
  }

  console.log(`${"=".repeat(50)}\n`);
}

main();

File: python/test_endpoints.py
"""
Script de teste para validar os endpoints da API.
Rode este script com o servidor ligado (python app.py) para ver quais endpoints
estao funcionando corretamente e quais estao com problemas.

Uso:
    python test_endpoints.py
"""

import urllib.request
import urllib.error
import json
import sys


BASE_URL = "http://localhost:5050"

GREEN = "\033[92m"
RED = "\033[91m"
YELLOW = "\033[93m"
RESET = "\033[0m"
BOLD = "\033[1m"


def fetch(path):
    """Faz GET para o endpoint e retorna o JSON parseado."""
    try:
        url = BASE_URL + path
        req = urllib.request.urlopen(url)
        return json.loads(req.read().decode()), None
    except urllib.error.HTTPError as e:
        return None, f"HTTP {e.code}: {e.reason}"
    except urllib.error.URLError as e:
        return None, f"Servidor inacessivel: {e.reason}"
    except Exception as e:
        return None, str(e)


def check(label, actual, expected):
    """Compara valor atual com esperado."""
    if actual == expected:
        print(f"  {GREEN}PASS{RESET} {label}")
        print(f"         Esperado: {expected}")
        return True
    else:
        print(f"  {RED}FAIL{RESET} {label}")
        print(f"         Esperado: {expected}")
        print(f"         Recebido: {actual}")
        return False


def check_approx(label, actual, expected, tolerance=0.01):
    """Compara floats com tolerancia."""
    if actual is not None and abs(actual - expected) < tolerance:
        print(f"  {GREEN}PASS{RESET} {label}")
        print(f"         Esperado: ~{expected}")
        return True
    else:
        print(f"  {RED}FAIL{RESET} {label}")
        print(f"         Esperado: ~{expected}")
        print(f"         Recebido: {actual}")
        return False


def test_endpoint(name, path, checks_fn):
    """Testa um endpoint: faz a request e roda as checagens."""
    print(f"\n{BOLD}--- {name} ---{RESET}")
    print(f"    GET {path}")

    data, error = fetch(path)

    if error:
        print(f"  {RED}ERRO{RESET} Nao foi possivel acessar o endpoint")
        print(f"         {error}")
        return 0, 1

    passed = 0
    failed = 0
    for result in checks_fn(data):
        if result:
            passed += 1
        else:
            failed += 1
    return passed, failed


def main():
    print(f"{BOLD}Verificando se o servidor esta rodando...{RESET}")
    try:
        urllib.request.urlopen(BASE_URL + "/api/revenue")
        print(f"{GREEN}Servidor OK!{RESET}")
    except urllib.error.HTTPError:
        print(f"{GREEN}Servidor OK!{RESET} (endpoints com erro — esperado)")
    except urllib.error.URLError as e:
        print(f"\n{RED}ERRO: O servidor nao esta acessivel em {BASE_URL}{RESET}")
        print("Certifique-se de que voce rodou: python app.py")
        print(f"Erro: {e.reason}")
        sys.exit(1)

    total_passed = 0
    total_failed = 0

    # =============================================
    # TESTE 1: Receita total (deve excluir cancelados)
    # =============================================
    p, f = test_endpoint(
        "Teste 1: Receita Total (excluindo pedidos cancelados)",
        "/api/revenue",
        lambda data: [
            check_approx("total_revenue", data.get("total_revenue"), 978.10),
        ]
    )
    total_passed += p
    total_failed += f

    # =============================================
    # TESTE 2: Receita por restaurante
    # =============================================
    p, f = test_endpoint(
        "Teste 2: Receita por Restaurante (excluindo cancelados)",
        "/api/revenue/restaurant",
        lambda data: [
            check_approx("Sushi Place", data.get("Sushi Place"), 384.90),
            check_approx("Burger House", data.get("Burger House"), 202.20),
        ]
    )
    total_passed += p
    total_failed += f

    # =============================================
    # TESTE 3: Top restaurantes
    # =============================================
    p, f = test_endpoint(
        "Teste 3: Top 3 Restaurantes (maior receita primeiro)",
        "/api/restaurants/top",
        lambda data: [
            check(
                "Ordem correta",
                [r["name"] for r in data.get("top_restaurants", [])],
                ["Sushi Place", "Pizza Roma", "Burger House"],
            ),
            check_approx(
                "Receita do #1 (Sushi Place)",
                data.get("top_restaurants", [{}])[0].get("revenue") if data.get("top_restaurants") else None,
                384.90,
            ),
        ]
    )
    total_passed += p
    total_failed += f

    # =============================================
    # TESTE 4: Produtos populares
    # =============================================
    p, f = test_endpoint(
        "Teste 4: Produtos Mais Populares",
        "/api/products/popular",
        lambda data: [
            check(
                "Produto #1 e Milkshake (5 unidades)",
                data.get("popular_products", [{}])[0].get("product") if data.get("popular_products") else None,
                "Milkshake",
            ),
            check(
                "Quantidade do #1",
                data.get("popular_products", [{}])[0].get("quantity") if data.get("popular_products") else None,
                5,
            ),
        ]
    )
    total_passed += p
    total_failed += f

    # =============================================
    # TESTE 5: Resumo geral
    # =============================================
    p, f = test_endpoint(
        "Teste 5: Resumo Geral",
        "/api/summary",
        lambda data: [
            check("total_orders (sem cancelados)", data.get("total_orders"), 12),
            check("cancelled_orders", data.get("cancelled_orders"), 3),
            check_approx("total_revenue", data.get("total_revenue"), 978.10),
            check("total_items_sold", data.get("total_items_sold"), 28),
        ]
    )
    total_passed += p
    total_failed += f

    # =============================================
    # RESULTADO FINAL
    # =============================================
    total = total_passed + total_failed
    print(f"\n{'='*50}")
    print(f"{BOLD}RESULTADO FINAL: {total_passed}/{total} testes passando{RESET}")

    if total_failed == 0:
        print(f"{GREEN}Todos os testes passaram! Parabens!{RESET}")
    else:
        print(f"{RED}{total_failed} teste(s) falhando{RESET}")
        print(f"\n{YELLOW}Dica: corrija os bugs um de cada vez e rode este script novamente.{RESET}")

    print(f"{'='*50}\n")


if __name__ == "__main__":
    main()
